dipy/reconst/weights_method.py

```diff
@@ -0,0 +1,248 @@
+
+import numpy as np
+
+MIN_POSITIVE_SIGNAL = 0.0001
+
+
+def simple_cutoff(residuals, log_residuals, pred_sig,
+                  design_matrix, leverages, C, cutoff):
+    """ Define outliers based on signal.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (np.abs(residuals) > + cutoff * C * HAT_factor)
+    return cond
+
+
+def two_eyes_cutoff(residuals, log_residuals, pred_sig,
+                    design_matrix, leverages, C, cutoff):
+    """ Two-eyes approach to define outliers, see [1]_.
+
+    References
+    ----------
+    .. [1] Collier, Q., Veraart, J., Jeurissen, B., den Dekker, A., Sijbers,
+       J., 2015. Iterative reweighted linear least squares for accurate, fast,
+       and robust estimation of diffusion magnetic resonance parameters: IRLLS
+       for Estimation of Diffusion MR Parameters. Magnetic Resonance in
+       Medicine 73, 2174-2184.
+    """
+    leverages[np.isclose(leverages, 1.0)] = 0.99  # avoids rare issues
+    HAT_factor = np.sqrt(1 - leverages)
+    cond = (residuals > + cutoff * C * HAT_factor) |\
+           (log_residuals < - cutoff * C * HAT_factor / pred_sig)
+    return cond
+
+
+def weights_method_wls_m_est(data, pred_sig, design_matrix, leverages,
+                             idx, total_idx, last_robust,
+                             m_est="gm", cutoff=3,
+                             outlier_condition_func=simple_cutoff):
+    """ Calculate M-estimator weights for WLS model.
+
+    Parameters
+    ----------
+    data : array
+        The measured signal.
+    pred_sig : array
+        The predicted signal, given a previous fit.
+        Has the same shape as data.
+    design_matrix : array (g, ...)
+        Design matrix holding the covariants used to solve for the
+        regression coefficients.
+    leverages : array
+        The leverages (diagonal of the 'hat matrix') of the fit.
+    idx : int
+        The current iteration number.
+    total_idx : int
+        The total number of iterations.
+    last_robust : array (int),
+        Array of 1s (inlier) and 0s (outlier).
+        Has the same shape as data.
+    m_est : which M-estimator weighting scheme to use. Currently,
+            'gm' (Geman-McClure) and 'cauchy' are provided.
+    cut_off : float
+        Outliers are defined by residuals > |cut_off * standard_deviation|
+
+    Notes
+    -----
+    Robust weights are calculated specifically for the WLS problem, i.e. the
+    usual form of the WLS problem is accounted for when defining these new
+    weights. On the second-to-last iteration, OLS is performed without
+    outliers. On the last iteration, WLS is performed without outliers.
+
+    """
+    # check if M-estimator is valid (defined in this function)
+    if m_est not in ["gm", "cauchy"]:
+        raise ValueError("unknown M-estimator choice")
+
+    # min 4 iters: (1) WLS (2) WLS with M-weights (3) clean OLS (3) clean WLS
```

Comment:
Not sure I understand this comment. These methods all require min 4 iterations?

Reply:
yes, because you implement robust IRLS for the WLS problem (the robust form of WLS) in this way, without 4 iterations, because of the flexible way IRLS implement works here. First, one must do WLS as usual, then one must fit with robust weights, then fitting with weights as either 1 or 0 based on outlier detection for OLS, then the final WLS step.